File: src/categories.ts
/**
 * Location categories — the "Find nearby" chips (Food, Groceries, Hotels, …),
 * modeled on Apple Maps. Each category carries a color + glyph for its map pin
 * and a set of OpenStreetMap tag filters used to query nearby places.
 */

export interface Category {
  id: string;
  label: string;
  color: string;
  /** Inner SVG (24×24, stroke-based) drawn inside the pin head. */
  glyph: string;
  /** OpenStreetMap [key, value] tag pairs to search for. */
  osm: [string, string][];
}

// A neutral pin for plain search results (Apple's red "dropped" pin).
export const DEFAULT_PIN_COLOR = '#FF3B30';

export const CATEGORIES: Category[] = [
  {
    id: 'food',
    label: 'Food',
    color: '#FF9500',
    glyph:
      '<path d="M6 3v5a2 2 0 0 0 4 0V3"/><path d="M8 8v13"/><path d="M16 3c-1.6 1-2.2 3-2.2 6s2.2 3 2.2 3v9"/>',
    osm: [
      ['amenity', 'restaurant'],
      ['amenity', 'fast_food'],
    ],
  },
  {
    id: 'coffee',
    label: 'Coffee',
    color: '#AC7F5E',
    glyph:
      '<path d="M5 8h11v4a5 5 0 0 1-5 5h-1a5 5 0 0 1-5-5V8Z"/><path d="M16 9h2.5a2.5 2.5 0 0 1 0 5H16"/><path d="M8 3v2"/><path d="M12 3v2"/>',
    osm: [['amenity', 'cafe']],
  },
  {
    id: 'groceries',
    label: 'Groceries',
    color: '#FFB340',
    glyph:
      '<circle cx="9" cy="20" r="1.4"/><circle cx="17" cy="20" r="1.4"/><path d="M3 4h2l2 11h11"/><path d="M7 8h13l-1.6 6"/>',
    osm: [
      ['shop', 'supermarket'],
      ['shop', 'convenience'],
      ['shop', 'grocery'],
    ],
  },
  {
    id: 'gas',
    label: 'Gas',
    color: '#5E5CE6',
    glyph:
      '<path d="M5 21V6a2 2 0 0 1 2-2h5a2 2 0 0 1 2 2v15"/><path d="M4 21h11"/><path d="M7 9h5"/><path d="M14 9l3 3v6a2 2 0 0 0 3 0V10l-3-3"/>',
    osm: [['amenity', 'fuel']],
  },
  {
    id: 'hotels',
    label: 'Hotels',
    color: '#BF5AF2',
    glyph:
      '<path d="M3 19V9"/><path d="M3 13h13a5 5 0 0 1 5 5v1"/><path d="M21 19v-2"/><circle cx="7.5" cy="11" r="1.6"/>',
    osm: [
      ['tourism', 'hotel'],
      ['tourism', 'motel'],
    ],
  },
  {
    id: 'shopping',
    label: 'Shopping',
    color: '#FF2D55',
    glyph: '<path d="M6 8h12l1 12H5L6 8Z"/><path d="M9 8a3 3 0 0 1 6 0"/>',
    osm: [
      ['shop', 'mall'],
      ['shop', 'department_store'],
      ['shop', 'clothes'],
    ],
  },
  {
    id: 'parks',
    label: 'Parks',
    color: '#34C759',
    glyph: '<path d="M12 3 6 12h3l-3 5h12l-3-5h3L12 3Z"/><path d="M12 17v4"/>',
    osm: [['leisure', 'park']],
  },
  {
    id: 'bars',
    label: 'Bars',
    color: '#64D2FF',
    glyph: '<path d="M5 4h14l-7 8-7-8Z"/><path d="M12 12v6"/><path d="M8 21h8"/>',
    osm: [
      ['amenity', 'bar'],
      ['amenity', 'pub'],
    ],
  },
  {
    id: 'surf',
    label: 'Surf',
    color: '#00C7BE',
    glyph: '<path d="M3 17c4 0 4-2 8-2s4 2 8 2"/><path d="M3 21c4 0 4-2 8-2s4 2 8 2"/><path d="M14 4c3 3 3 7 0 11"/>',
    osm: [
      ['sport', 'surfing'],
      ['natural', 'beach'],
    ],
  },
  {
    id: 'ski',
    label: 'Ski',
    color: '#5AC8FA',
    glyph: '<path d="M4 20l16-6"/><path d="M6 19l14-5"/><circle cx="16" cy="5" r="1.6"/><path d="M14 8l2 3 3 1"/>',
    osm: [
      ['sport', 'skiing'],
      ['landuse', 'winter_sports'],
    ],
  },
  {
    id: 'climb',
    label: 'Climb',
    color: '#AF52DE',
    glyph: '<circle cx="14" cy="5" r="1.6"/><path d="M13 8l-4 3 3 3-2 6"/><path d="M12 14l5 2 3-2"/><path d="M9 11l-4 1"/>',
    osm: [['sport', 'climbing']],
  },
  {
    id: 'golf',
    label: 'Golf',
    color: '#30D158',
    glyph: '<path d="M11 3v14"/><path d="M11 5l6 2-6 2"/><path d="M6 21c1-1.5 3-2 5-2s4 .5 5 2"/>',
    osm: [['leisure', 'golf_course']],
  },
  {
    id: 'camp',
    label: 'Camp',
    color: '#FF9500',
    glyph: '<path d="M12 4 3 20h18L12 4Z"/><path d="M12 4v16"/>',
    osm: [
      ['tourism', 'camp_site'],
      ['tourism', 'wilderness_hut'],
    ],
  },
];

export function categoryById(id: string | undefined): Category | undefined {
  return id ? CATEGORIES.find((c) => c.id === id) : undefined;
}

File: phantom_eye/packages/pe_domain/lib/src/places/categories.dart
/// One OSM `key=value` tag pair used to query a category.
final class OsmTag {
  const OsmTag(this.key, this.value);
  final String key;
  final String value;

  @override
  String toString() => '$key=$value';
}

/// A searchable place category, matching the chips in the legacy UI.
///
/// `colorHex` is kept as a string rather than a Flutter `Color` so this stays
/// in the Flutter-free domain layer; the UI package converts it.
final class Category {
  const Category({
    required this.id,
    required this.label,
    required this.colorHex,
    required this.osmTags,
    required this.keywords,
  });

  final String id;
  final String label;
  final String colorHex;

  /// Tag pairs to query. Each becomes one node clause and one way clause.
  final List<OsmTag> osmTags;

  /// Free-text terms that should resolve to this category.
  final List<String> keywords;
}

/// Fallback pin colour for results with no category.
///
/// Ported from `DEFAULT_PIN_COLOR` in `src/categories.ts`.
const String kDefaultPinColorHex = '#FF3B30';

/// Filler words stripped before category matching, so "best coffee near me"
/// still resolves to Coffee.
///
/// Ported from `FILLER` in `src/categories.ts:192`. Order matters inside the
/// alternation only insofar as the regex is applied globally.
final RegExp _filler = RegExp(
  r'\b(near ?me|near ?by|around ?me|close ?by|closest|nearest|places?|spots?'
  r'|good|best|cheap|open|the|a|some)\b',
);

/// Everything outside this set is replaced with a space.
///
/// Note this **strips digits**, so "76 gas" reduces to "gas" — preserved from
/// the original, where it still matches the Gas category via the remaining
/// word. Accented Latin letters are kept so "café" survives.
final RegExp _nonLetters = RegExp(r'[^a-zà-ÿ\s]');

final RegExp _whitespace = RegExp(r'\s+');

/// The 13 place categories, in declaration order — which is also match
/// precedence, since matching is first-match-wins.
///
/// Generated from `CATEGORIES` in `src/categories.ts` rather than transcribed
/// by hand, so ids, labels, colours, OSM tag pairs and keyword lists cannot
/// drift from the original. The glyph SVGs are deliberately omitted: they
/// belong to the UI layer, not here.
const List<Category> kCategories = [
  Category(
    id: 'food',
    label: 'Food',
    colorHex: '#FF9500',
    osmTags: [OsmTag('amenity', 'restaurant'), OsmTag('amenity', 'fast_food')],
    keywords: [
      'food',
      'restaurant',
      'restaurants',
      'dinner',
      'lunch',
      'brunch',
      'dining',
      'eat',
      'meal',
      'takeout',
      'diner',
      'pizza',
      'burger',
      'burgers',
      'sushi',
      'tacos',
      'mexican',
      'italian',
      'chinese',
      'thai',
      'indian',
      'sandwich',
      'bbq',
      'steakhouse',
    ],
  ),
  Category(
    id: 'coffee',
    label: 'Coffee',
    colorHex: '#AC7F5E',
    osmTags: [OsmTag('amenity', 'cafe')],
    keywords: [
      'coffee',
      'cafe',
      'café',
      'espresso',
      'latte',
      'cappuccino',
      'breakfast',
    ],
  ),
  Category(
    id: 'groceries',
    label: 'Groceries',
    colorHex: '#FFB340',
    osmTags: [
      OsmTag('shop', 'supermarket'),
      OsmTag('shop', 'convenience'),
      OsmTag('shop', 'grocery'),
    ],
    keywords: ['groceries', 'grocery', 'supermarket', 'market', 'food store'],
  ),
  Category(
    id: 'gas',
    label: 'Gas',
    colorHex: '#5E5CE6',
    osmTags: [OsmTag('amenity', 'fuel')],
    keywords: ['gas', 'fuel', 'petrol', 'gasoline', 'gas station'],
  ),
  Category(
    id: 'hotels',
    label: 'Hotels',
    colorHex: '#BF5AF2',
    osmTags: [OsmTag('tourism', 'hotel'), OsmTag('tourism', 'motel')],
    keywords: [
      'hotel',
      'hotels',
      'motel',
      'lodging',
      'lodge',
      'inn',
      'stay',
      'accommodation',
    ],
  ),
  Category(
    id: 'shopping',
    label: 'Shopping',
    colorHex: '#FF2D55',
    osmTags: [
      OsmTag('shop', 'mall'),
      OsmTag('shop', 'department_store'),
      OsmTag('shop', 'clothes'),
    ],
    keywords: [
      'shopping',
      'mall',
      'shops',
      'store',
      'stores',
      'outlet',
      'clothes',
    ],
  ),
  Category(
    id: 'parks',
    label: 'Parks',
    colorHex: '#34C759',
    osmTags: [OsmTag('leisure', 'park')],
    keywords: ['park', 'parks', 'playground', 'green space'],
  ),
  Category(
    id: 'bars',
    label: 'Bars',
    colorHex: '#64D2FF',
    osmTags: [OsmTag('amenity', 'bar'), OsmTag('amenity', 'pub')],
    keywords: [
      'bar',
      'bars',
      'pub',
      'pubs',
      'drinks',
      'nightlife',
      'brewery',
      'tavern',
      'cocktails',
    ],
  ),
  Category(
    id: 'surf',
    label: 'Surf',
    colorHex: '#00C7BE',
    osmTags: [OsmTag('sport', 'surfing'), OsmTag('natural', 'beach')],
    keywords: ['surf', 'surfing', 'waves', 'beach', 'beaches', 'surf spot'],
  ),
  Category(
    id: 'ski',
    label: 'Ski',
    colorHex: '#5AC8FA',
    osmTags: [OsmTag('sport', 'skiing'), OsmTag('landuse', 'winter_sports')],
    keywords: [
      'ski',
      'skiing',
      'snowboard',
      'snowboarding',
      'slopes',
      'ski resort',
    ],
  ),
  Category(
    id: 'climb',
    label: 'Climb',
    colorHex: '#AF52DE',
    osmTags: [OsmTag('sport', 'climbing')],
    keywords: ['climb', 'climbing', 'bouldering', 'rock climbing'],
  ),
  Category(
    id: 'golf',
    label: 'Golf',
    colorHex: '#30D158',
    osmTags: [OsmTag('leisure', 'golf_course')],
    keywords: ['golf', 'golf course', 'driving range'],
  ),
  Category(
    id: 'camp',
    label: 'Camp',
    colorHex: '#FF9500',
    osmTags: [
      OsmTag('tourism', 'camp_site'),
      OsmTag('tourism', 'wilderness_hut'),
    ],
    keywords: ['camp', 'camping', 'campground', 'campsite', 'campsites'],
  ),
];

/// Look up a category by id. Null id or unknown id yields null.
///
/// Ported from `categoryById` in `src/categories.ts:188`.
Category? categoryById(String? id) {
  if (id == null || id.isEmpty) return null;
  for (final c in kCategories) {
    if (c.id == id) return c;
  }
  return null;
}

/// Normalise a free-text query for category matching.
///
/// Exposed for testing so the stripping stages can be checked directly.
String normalizeCategoryQuery(String query) => query
    .toLowerCase()
    .replaceAll(_filler, ' ')
    .replaceAll(_nonLetters, ' ')
    .replaceAll(_whitespace, ' ')
    .trim();

/// Map a free-text query to a category, so typing "dinner" or "gas" behaves
/// like tapping the matching chip.
///
/// Returns null when the query looks like a specific place name, which the
/// caller then hands to geocoding instead.
///
/// Ported from `matchCategory` in `src/categories.ts:199`. Two passes, in this
/// order:
///  1. **whole phrase** — the normalised query equals a label or is one of a
///     category's keywords. This is what lets multi-word keywords such as
///     `gas station` and `food store` match at all.
///  2. **tokens** — only for queries of three words or fewer, any single token
///     matching a label or keyword. The length cap keeps a long place name
///     containing an incidental word (say "The Coffee Manufactory on Bryant")
///     from being hijacked into a category search.
///
/// Both passes scan [kCategories] in order and return the first hit.
Category? matchCategory(String query) {
  final q = normalizeCategoryQuery(query);
  if (q.isEmpty) return null;

  for (final c in kCategories) {
    if (c.label.toLowerCase() == q || c.keywords.contains(q)) return c;
  }

  final tokens = q.split(' ');
  if (tokens.length <= 3) {
    for (final c in kCategories) {
      final set = {c.label.toLowerCase(), ...c.keywords};
      if (tokens.any(set.contains)) return c;
    }
  }
  return null;
}
